core/src/main/scala/zio/prelude/ops/package.scala

```diff
@@ -0,0 +1,40 @@
+package zio.prelude
+
+import zio.prelude.fx.ZPure
+
+package object ops {
+
+  implicit class EnrichZPure(z: ZPure.type) {
+
+    def whenCase[W, S1, R, E, A, B](a: => A)(
+      pf: PartialFunction[A, ZPure[W, S1, S1, R, E, B]]
+    ): ZPure[W, S1, S1, R, E, Option[B]] =
+      ZPure.suspend(pf.lift(a).map(_.map(Some(_))).getOrElse(ZPure.succeed(None)))
+
+    def unless[W, S1, R, E, A](condition: => Boolean)(
+      pure: => ZPure[W, S1, S1, R, E, A]
+    ): ZPure[W, S1, S1, R, E, Option[A]] =
+      when[W, S1, R, E, A](!condition)(pure)
+
+    def when[W, S1, R, E, A](condition: => Boolean)(
+      pure: => ZPure[W, S1, S1, R, E, A]
+    ): ZPure[W, S1, S1, R, E, Option[A]] =
+      ZPure.suspend(if (condition) pure.map(Some(_)) else ZPure.succeed(None))
+
+    def forEachDiscard[W, S1, R, E, A](as: Iterable[A])(
+      f: A => ZPure[W, S1, S1, R, E, Any]
+    ): ZPure[W, S1, S1, R, E, Unit] =
+      ZPure.suspend {
+        val iterator = as.iterator
+        if (iterator.hasNext) {
+          var result: ZPure[W, S1, S1, R, E, Any] = f(iterator.next())
+          while (iterator.hasNext)
+            result = result *> f(iterator.next())
+          result.unit
+        } else {
+          ZPure.unit
+        }
+      }
+  }
```

Comment:
My gut feeling is that the usage of by-name arguments (except `pure: => ZPure[W, S1, S1, R, E, A]` since we probably want that to be by-name) and `ZPure.suspend(...)` here will have a rather noticeable performance impact. Would something like this work or do we strictly need the methods to return lazily constructed `ZPure`s?

```scala
    def whenCase[W, S1, R, E, A, B](a: A)(
      pf: PartialFunction[A, ZPure[W, S1, S1, R, E, B]]
    ): ZPure[W, S1, S1, R, E, Option[B]] =
      pf.lift(a).map(_.map(Some(_))).getOrElse(ZPure.succeed(None))

    def unless[W, S1, R, E, A](condition: Boolean)(
      pure: => ZPure[W, S1, S1, R, E, A]
    ): ZPure[W, S1, S1, R, E, Option[A]] =
      when[W, S1, R, E, A](!condition)(pure)

    def when[W, S1, R, E, A](condition: Boolean)(
      pure: => ZPure[W, S1, S1, R, E, A]
    ): ZPure[W, S1, S1, R, E, Option[A]] =
      if (condition) pure.map(Some(_)) else ZPure.succeed(None)

    def forEachDiscard[W, S1, R, E, A](as: Iterable[A])(
      f: A => ZPure[W, S1, S1, R, E, Any]
    ): ZPure[W, S1, S1, R, E, Unit] = {
        val iterator = as.iterator
        if (iterator.hasNext) {
          var result: ZPure[W, S1, S1, R, E, Any] = f(iterator.next())
          while (iterator.hasNext)
            result = result *> f(iterator.next())
          result.unit
        } else {
          ZPure.unit
        }
      }
  }
```

Reply:
if we make our own `when`/`whenCase`, maybe we could return `Unit` as it used to be, so that we don't need to do `.unit` everywhere 😄

Reply:
![Screenshot 2023-02-27 at 16 29 27](https://user-images.githubusercontent.com/43631/221606495-11499d68-3715-4927-98a9-f2fb61bcb11c.png)